proof/courier.go

```diff
@@ -90,20 +55,123 @@ type CourierHarness interface {
 // receiver can use this to fetch a proof from the sender.
 //
 // TODO(roasbeef): FileSystemCourier, RpcCourier
-type Courier[Addr any] interface {
+type Courier interface {
 	// DeliverProof attempts to delivery a proof to the receiver, using the
 	// information in the Addr type.
-	DeliverProof(context.Context, Addr, *AnnotatedProof) error
+	DeliverProof(context.Context, Recipient, *AnnotatedProof) error
 
 	// ReceiveProof attempts to obtain a proof as identified by the passed
 	// locator from the source encapsulated within the specified address.
-	ReceiveProof(context.Context, Addr, Locator) (*AnnotatedProof, error)
+	ReceiveProof(context.Context, Recipient, Locator) (*AnnotatedProof,
+		error)
 
 	// SetSubscribers sets the set of subscribers that will be notified
 	// of proof courier related events.
 	SetSubscribers(map[uint64]*fn.EventReceiver[fn.Event])
 }
 
+// CourierAddr is a fully validated courier address (including protocol specific
+// validation).
+type CourierAddr interface {
+	// Url returns the url.URL representation of the courier address.
+	Url() *url.URL
+
+	// NewCourier generates a new courier service handle.
+	NewCourier(ctx context.Context, cfg *CourierCfg) (Courier, error)
```

Comment:
Similarly here, will the config actually be universal? What extra information does an addr need, that isn't in the addr itself to create a courier instance?

Reply:
Extra info are things such as:

* TLS certificate
* Backoff procedure parameters
* timeout

Reply:
I think TLS certificates we should just ignore with `tlsConfig = &tls.Config{InsecureSkipVerify: true}`  (and assume that most universes will have a valid, public certificate CA, and if not, we shouldn't fail on that). Since the proofs aren't super sensitive, we don't need to care about transport security that much.